src/include/processor/operator/recursive_extend/bfs_state.h

```diff
@@ -30,18 +83,85 @@ class TargetDstNodes {
 
     inline uint64_t getNumNodes() const { return numNodes; }
 
+    inline frontier::node_id_set_t getNodeIDs() const { return nodeIDs; }
+
 private:
     uint64_t numNodes;
     frontier::node_id_set_t nodeIDs;
     std::unordered_set<common::table_id_t> tableIDFilter;
 };
 
-class BaseBFSState {
+struct BaseBFSMorsel;
+
+/**
+ * An SSSPSharedState is a unit of work for the nTkS scheduling policy. It is *ONLY* used for the
+ * particular case of SHORTEST_PATH | SINGLE_LABEL | NO_PATH_TRACKING. An SSSPSharedState is *NOT*
+ * exclusive to a thread and any thread can pick up BFS extension or Writing Path Length.
+ * A shared_ptr is maintained by the BaseBFSMorsel and a morsel of work is fetched using this ptr.
+ */
+struct SSSPSharedState {
+public:
+    SSSPSharedState(uint64_t upperBound_, uint64_t lowerBound_, uint64_t maxNodeOffset_)
+        : mutex{std::mutex()}, ssspLocalState{EXTEND_IN_PROGRESS}, currentLevel{0u},
+          nextScanStartIdx{0u}, numVisitedNodes{0u}, visitedNodes{std::vector<uint8_t>(
+                                                         maxNodeOffset_ + 1, NOT_VISITED)},
+          pathLength{std::vector<uint8_t>(maxNodeOffset_ + 1, 0u)},
+          bfsLevelNodeOffsets{std::vector<common::offset_t>()}, srcOffset{0u},
+          maxOffset{maxNodeOffset_}, upperBound{upperBound_}, lowerBound{lowerBound_},
+          numThreadsActiveOnMorsel{0u}, nextDstScanStartIdx{0u}, inputFTableTupleIdx{0u} {}
+
+    void reset(TargetDstNodes* targetDstNodes);
+
+    bool getBFSMorsel(std::unique_ptr<BaseBFSMorsel>& bfsMorsel, SSSPLocalState& state);
+
+    bool hasWork() const;
+
+    bool finishBFSMorsel(std::unique_ptr<BaseBFSMorsel>& bfsMorsel);
+
+    // If BFS has completed.
+    bool isComplete(uint64_t numDstNodesToVisit);
+    // Mark src as visited.
+    void markSrc(bool isSrcDestination);
+
+    void moveNextLevelAsCurrentLevel();
+
+    std::pair<uint64_t, int64_t> getDstPathLengthMorsel();
+
+public:
+    std::mutex mutex;
+    SSSPLocalState ssspLocalState;
+    // Level state
+    uint8_t currentLevel;
+
+    uint64_t nextScanStartIdx;
+
+    // Visited state
+    uint64_t numVisitedNodes;
+    std::vector<uint8_t> visitedNodes;
+    std::vector<uint8_t> pathLength;
+    std::vector<common::offset_t> bfsLevelNodeOffsets;
+    // Offset of src node.
+    common::offset_t srcOffset;
+    // Maximum offset of dst nodes.
+    common::offset_t maxOffset;
+    uint64_t upperBound;
+    uint64_t lowerBound;
+    uint32_t numThreadsActiveOnMorsel;
+    uint64_t nextDstScanStartIdx;
+    uint64_t inputFTableTupleIdx;
+};
+
+struct BaseBFSMorsel {
```

Comment:
Once you finish migrating all type of recursive join. One of your `BaseBFSMorsel` and `SSSPSharedState` should be simplified. Most likely your `BaseBFSMorsel` is the one that only stores a pointer to `SSSPSharedState` plus a range that is locally visible.

Reply:
yeah once all shortest, shortest and var len are added